olive/passes/onnx/extract_adapters.py

```diff
@@ -0,0 +1,317 @@
+            weights = packed_weights
+
+        # update the model with the changes
+        dag.update()
+
+        # save the weights
+        # TODO(jambayk): Consider other methods for saving the weights
+        # safetensors is an option but it is not available on ARM64 Windows
+        weights_path = Path(output_model_path).parent / "adapter_weights.npz"
+        np.savez(weights_path, **weights)
+
+        # save the model
+        output_model = model_proto_to_olive_model(
+            dag.model,
+            output_model_path,
+            external_data_config={"save_as_external_data": True, "all_tensors_to_one_file": True},
+            external_initializers_file_name=weights_path.name if not config["make_inputs"] else None,
+            constant_inputs_file_name=weights_path.name if config["make_inputs"] else None,
+        )
+        output_model.model_attributes = deepcopy(model.model_attributes)
+        # save information about the weights in the model attributes
+        weights_info = {name: [list(value.shape), str(value.dtype)] for name, value in weights.items()}
+        if not config["make_inputs"]:
+            output_model.model_attributes["external_initializers"] = weights_info
+        else:
+            output_model.model_attributes["constant_inputs"] = weights_info
+            if config["pack_inputs"]:
+                output_model.model_attributes["packed_inputs"] = packings
+        return output_model
+
+    @staticmethod
+    def _get_lora_name_patterns(lora_modules: List[str]) -> List[str]:
+        """Get the node name patterns for lora modules."""
+        return [f".*[./]{key}[./]{name}[./]MatMul$" for key in lora_modules for name in ["default", "default_1"]]
+
+    @staticmethod
+    def _create_new_weight_name(old_name: str) -> str:
+        """Create new weight name based on old name.
+
+        The new weight name is of the form model.layers.0.self_attn.q_proj.lora_A.quant.weight
+        """
+        weight_name = old_name[1:] if old_name.startswith("/") else old_name
+        return (
+            weight_name.replace("/", ".")
+            .replace("default.", "lora_A.")
+            .replace("default_1.", "lora_B.")
+            .replace("MatMul", "weight")
+        )
+
+    @staticmethod
+    def _copy_initializer(old_initializer: onnx.TensorProto, new_name: str) -> onnx.TensorProto:
+        """Copy initializer with a new name and dummy external data location."""
+        from onnx.external_data_helper import set_external_data
+
+        # create a new initializer
+        new_initializer = onnx.TensorProto()
+        # copy the old initializer
+        new_initializer.CopyFrom(old_initializer)
+        # set the new name
+        new_initializer.name = new_name
+        # raw_data is required for set_external_data
+        if not new_initializer.HasField("raw_data"):
+            new_initializer.raw_data = b""
+        set_external_data(new_initializer, location="dummy-location.bin")
```

Comment:
Will this external data be stored under the path user running olive?
Seems it is `dummy` stuff? Where will we remove it?

Reply:
This is actually just a place holder. Since the raw_data field is cleared, during model save, it doesn't save anything at this location. https://github.com/onnx/onnx/blob/main/onnx/external_data_helper.py#L304

This is why the description of this PR says the onnx model produced by extract as initializers is invalid because it points to non-existent file.
This is okay since we only intend to use this model using ort inference session.